controllers/collector_controller.go

```diff
@@ -0,0 +1,233 @@
+			return ctrl.Result{}, fmt.Errorf("error when adding finalizer: %v", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Check if Secret exists and requeue when not found
+	var sec corev1.Secret
+	if err := r.Get(ctx, client.ObjectKey{Namespace: co.Namespace, Name: co.Spec.FluentBitConfigName}, &sec); err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{Requeue: true}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Install RBAC resources for the filter plugin kubernetes
+	var rbacObj, saObj, bindingObj client.Object
+	rbacObj, saObj, bindingObj = operator.MakeRBACObjects(co.Name, co.Namespace, "collector", co.Spec.RBACRules)
+	// Set ServiceAccount's owner to this fluentbit
+	if err := ctrl.SetControllerReference(&co, saObj, r.Scheme); err != nil {
+		return ctrl.Result{}, err
+	}
+	if err := r.Create(ctx, rbacObj); err != nil && !errors.IsAlreadyExists(err) {
+		return ctrl.Result{}, err
+	}
+	if err := r.Create(ctx, saObj); err != nil && !errors.IsAlreadyExists(err) {
+		return ctrl.Result{}, err
+	}
+	if err := r.Create(ctx, bindingObj); err != nil && !errors.IsAlreadyExists(err) {
+		return ctrl.Result{}, err
+	}
+
+	// Deploy Fluent Bit Statefuset
+	sts := operator.MakefbStatefuset(co)
+	if err := ctrl.SetControllerReference(&co, &sts, r.Scheme); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if _, err := controllerutil.CreateOrPatch(ctx, r.Client, &sts, r.mutate(&sts, co)); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Deploy collector Service
+	if !co.Spec.DisableService {
+		svc := operator.MakeCollecotrService(co)
+		if err := ctrl.SetControllerReference(&co, &svc, r.Scheme); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if _, err := controllerutil.CreateOrPatch(ctx, r.Client, &svc, r.mutate(&svc, co)); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *CollectorReconciler) mutate(obj client.Object, co fluentbitv1alpha2.Collector) controllerutil.MutateFn {
+	switch o := obj.(type) {
+	case *appsv1.StatefulSet:
+		expected := operator.MakefbStatefuset(co)
+
+		return func() error {
+			o.Labels = expected.Labels
+			o.Annotations = expected.Annotations
+			o.Spec = expected.Spec
+			o.SetOwnerReferences(nil)
+			if err := ctrl.SetControllerReference(&co, o, r.Scheme); err != nil {
+				return err
+			}
+			return nil
+		}
+	case *corev1.Service:
+		expected := operator.MakeCollecotrService(co)
+
+		return func() error {
+			o.Labels = expected.Labels
+			o.Spec.Selector = expected.Spec.Selector
+			o.Spec.Ports = expected.Spec.Ports
+			o.SetOwnerReferences(nil)
+			if err := ctrl.SetControllerReference(&co, o, r.Scheme); err != nil {
+				return err
+			}
+			return nil
+		}
+
+	default:
+	}
+
+	return nil
+}
+
+func (r *CollectorReconciler) delete(ctx context.Context, co *fluentbitv1alpha2.Collector) error {
+	var sa corev1.ServiceAccount
+	err := r.Get(ctx, client.ObjectKey{Namespace: co.Namespace, Name: co.Name}, &sa)
```

Comment:
No need to get it before deleting, just delete it directly.

Why does not delete the `clusterrole` and `clusterrolebinding`?